README.md
## Introduction

 **GChart** is a Haskell wrapper around [Google Chart API].

[Google Chart API]: http://code.google.com/apis/chart/image_charts.html

There is a library on Hackage called [GoogleChart] which provides another
wrapper, however it has not been updated in a long while. GChart improves upon
that effort by trying to design a more elegant API, and supporting more chart
types and features.

[GoogleChart]: http://hackage.haskell.org/packages/archive/GoogleChart/0.2/doc/html/Graphics-Google-Chart.html

## Installation

    cabal install hs-gchart

## Documentation

* [API functions reference](http://hackage.haskell.org/package/hs-gchart)

* [Haskell data types reference](http://hackage.haskell.org/packages/archive/hs-gchart/0.4/doc/html/Graphics-GChart-Types.html)

* [Google Chart API]

## Getting Started

These examples below are available in `examples/Examples.hs` in the source
tarball. Look at the source file for more Some examples are taken from [this
article](http://24ways.org/2007/tracking-christmas-cheer-with-google-charts)

For examples 1 and 2, the following code is common

    dataSeries1 :: [Int]
    dataSeries1 = [10,20,8,25,5,3,15,9,5]

    labelSeries1 = ["Egg nog",
                    "Christmas Ham",
                    "Milk (not including egg nog)",
                    "Cookies",
                    "Roasted Chestnuts",
                    "Chocolate",
                    "Various Other Beverages",
                    "Various Other Foods",
                    "Snacks"]

### Example 1 : Pie Chart

The code below

     christmasPie = getChartUrl $ do setChartSize 600 300
                                     setDataEncoding simple
                                     setChartType Pie
                                     addChartData dataSeries1
                                     setChartTitle "Food and Drink Consumed Christmas 2007"
                                     setLabels labelSeries1
                                     setColors ["00AF33",
                                                "4BB74C",
                                                "EE2C2C",
                                                "CC3232",
                                                "33FF33",
                                                "66FF66",
                                                "9AFF9A",
                                                "C1FFC1",
                                                "CCFFCC" ]

Generates the following chart.

![Generated Pie Chart](http://chart.apis.google.com/chart?cht=p&chs=600x300&chd=s:KUIZFDPJF&chtt=Food+and+Drink+Consumed+Christmas+2007&chco=00AF33,4BB74C,EE2C2C,CC3232,33FF33,66FF66,9AFF9A,C1FFC1,CCFFCC&chl=Egg+nog|Christmas+Ham|Milk+%28not+including+egg+nog%29|Cookies|Roasted+Chestnuts|Chocolate|Various+Other+Beverages|Various+Other+Foods|Snacks)

### Example 2 : Bar Chart

The code below

    barGraph = getChartUrl $ do setChartSize 600 300
                                setDataEncoding simple
                                setChartType BarHorizontalGrouped
                                addChartData dataSeries1
                                setChartTitle "Food and Drink Consumed Christmas 2007"
                                addAxis $ makeAxis {  axisType = AxisBottom }
                                addAxis $ makeAxis {  axisType = AxisLeft,
                                                      axisLabels = Just labelSeries1 }
                                addColor "00AF33"

Generates the following chart

![Generated Bar Chart](http://chart.apis.google.com/chart?cht=bhg&chs=600x300&chd=s:KUIZFDPJF&chtt=Food+and+Drink+Consumed+Christmas+2007&chco=00AF33&chxt=x,y&chxl=1:|Egg+nog|Christmas+Ham|Milk+%28not+including+egg+nog%29|Cookies|Roasted+Chestnuts|Chocolate|Various+Other+Beverages|Various+Other+Foods|Snacks)


### Example 3 : Line XY Chart 1

The code below

    linexyGraph1 =
        getChartUrl $ do setChartSize 800 300
                         setChartType LineXY
                         setDataEncoding text
                         setChartTitle "Projected Christmas Cheer for 2007"
                         setGrid $ makeGrid { xAxisStep = 3.333,
                                              yAxisStep = 10,
                                              lineSegmentLength = Just 1,
                                              blankSegmentLength = Just 3 }
                         addAxis $ makeAxis { axisType = AxisLeft,
                                              axisRange = Just $ Range (0,100) (Just 50) }
                         addAxis $ makeAxis { axisType = AxisBottom,
                                              axisLabels = Just $ ["Dec 1st"] ++ blanks 4 ++ ["6th"] ++ blanks 18 ++ ["25th","26th"] ++ blanks 4 ++ ["Dec 31st"] }
                         addChartDataXY dataSeries2

    dataSeries2 :: [(Float,Float)]
    dataSeries2 = [(0,0),(100,100)]

    blanks x = take x $ repeat ""

Generates the following chart

![Line Graph](http://chart.apis.google.com/chart?cht=lxy&chs=800x300&chd=t:0,100|0,100&chtt=Projected+Christmas+Cheer+for+2007&chxt=y,x&chxl=1:|Dec+1st|||||6th|||||||||||||||||||25th|26th|||||Dec+31st&chxr=0,0.0,100.0,50.0&chg=3.333,10.0,1.0,3.0)

### Example 4 : Line XY Chart 2

The code below

    linexyGraph2 =
        getChartUrl $ do setChartSize 800 300
                         setChartType LineXY
                         setDataEncoding text
                         setChartTitle "Projected Christmas Cheer for 2007"

                         setGrid $ makeGrid { xAxisStep = 3.333,
                                              yAxisStep = 10,
                                              lineSegmentLength = Just 1,
                                              blankSegmentLength = Just 3 }

                         addAxis $ makeAxis { axisType = AxisLeft,
                                              axisRange = Just $ Range (0,100) (Just 50) }
                         addAxis $ makeAxis { axisType = AxisBottom,
                                              axisLabels = Just $ ["Dec 1st"] ++ blanks 4 ++ ["6th"] ++ blanks 18 ++ ["25th","26th"] ++ blanks 4 ++ ["Dec 31st"] }

                         addChartDataXY dataSeries3
                         addColor "458B00"

                         addChartDataXY dataSeries4
                         addColor "CD2626"

                         setLegend $ legendWithPosition ["2006","2007"] LegendRight

    dataSeries3 :: [(Float,Float)]
    dataSeries3 = zip [0,16.7,23.3,33.3,60,76.7,83.3,86.7,93.3,96.7,100] [30,45,20,50,15,80,60,70,40,55,80]

    dataSeries4 :: [(Float,Float)]
    dataSeries4 = zip [0,10,16.7,26.7,33.3] [50,10,30,55,60]

    blanks x = take x $ repeat ""

Generates the following chart

![Line Graph](http://chart.apis.google.com/chart?cht=lxy&chs=800x300&chd=t:0,16.7,23.3,33.3,60,76.7,83.3,86.7,93.3,96.7,100|30,45,20,50,15,80,60,70,40,55,80|0,10,16.7,26.7,33.3|50,10,30,55,60&chtt=Projected+Christmas+Cheer+for+2007&chco=458B00,CD2626&chdl=2006|2007&chdlp=r&chxt=y,x&chxl=1:|Dec+1st|||||6th|||||||||||||||||||25th|26th|||||Dec+31st&chxr=0,0.0,100.0,50.0&chg=3.333,10.0,1.0,3.0)

### Example 5 : Scatter Plot with Shape Markers

The code below

    scatterPlotWithMarkers = getChartUrl $ do setChartSize 200 125
                                              setChartType ScatterPlot
                                              setDataEncoding simple
                                              addChartDataXY dataSeries5
                                              addAxis $ makeAxis { axisType = AxisBottom,
                                                                   axisLabels = Just $ blanks 1 ++ ["1","2","3","4","5"] }
                                              addAxis $ makeAxis { axisType = AxisLeft,
                                                                   axisLabels = Just $ blanks 1 ++ ["50","100"] }
                                              setGrid $ makeGrid { xAxisStep = 20, yAxisStep = 25 }

                                              addShapeMarker $ makeShapeMarker { shapeType  = ShapeSquare
                                                                               , shapeColor = "ff0000"
                                                                               , shapeSize  = 10 }

    -- Reverse engineering sample data from webpage
    dataSeries5 :: [(Int,Int)]
    dataSeries5 = zip xseries yseries where
                 xseries = map encSimpleReverse "984sttvuvkQIBLKNCAIipr3z9"
                 yseries = map encSimpleReverse "DEJPgq0uov17_zwopQOD"

    encSimpleReverse :: Char -> Int
    encSimpleReverse c | ord c >= ord 'A' && ord c <= ord 'Z' = (ord c - ord 'A')
                       | ord c >= ord 'a' && ord c <= ord 'z' = 26 + (ord c - ord 'a')
                       | ord c >= ord '0' && ord c <= ord '9' = 52 + (ord c - ord '0')
                       | otherwise = -1

Generates the following chart

![Scatter Plot with Shape Markers](http://chart.apis.google.com/chart?cht=s&chs=200x125&chd=s:984sttvuvkQIBLKNCAIi,DEJPgq0uov17_zwopQOD&chxt=x,y&chxl=0:||1|2|3|4|5|1:||50|100&chg=20.0,25.0&chm=s,ff0000,-1,-1,10)


CHANGELOG.md
## 0.4.2

* Brought up-to-date with recent GHCs


Graphics/GChart.hs
{-# LANGUAGE TypeSynonymInstances, NoMonomorphismRestriction #-}
{-|

Import this module to generate charts using the Google Chart API.

For more information about the Google Chart API, refer to

- Chart API Intro <http://code.google.com/apis/chart/image_charts.html>

- Getting Started <http://code.google.com/apis/chart/docs/making_charts.html>

For documentation full Haskell data model, refer to "Graphics.GChart.Types".

Here is an example to use the functions in the module to generate a chart URL :

@
generatePieChart = getChartUrl $ do setChartSize 640 400
                                 setChartType Pie
                                 setChartTitle \"Test\"
                                 addChartData  ([1,2,3,4,5]::[Int])
                                 addColor \"FF0000\"
                                 setLegend $ legend [\"t1\",\"t2\", \"t3\",\"t4\",\"t5\"]
                                 setLabels $ [\"Test 1\", \"Test 2\", \"Test 3\", \"Test 4\", \"Test 5\"]
@

For examples, refer to @Examples.hs@ in the source tarball, or download it
directly from Github :
<http://github.com/deepakjois/hs-gchart/blob/master/examples/Examples.hs>.


The module constists of:

- Smart Constructors - to make it convenient to construct data types

- Functions to set chart data

- Functions to retrieve chart data in form of URL or Haskell data type

-}
module Graphics.GChart (
  module Graphics.GChart.Types,

  -- * Smart Constructors
   solid                ,
   legend               ,
   legendWithPosition   ,
   makeAxis             ,
   makeGrid             ,
   simple               ,
   text                 ,
   extended             ,
   automatic            ,
   automaticWithSpacing ,
   barwidth             ,
   barwidthspacing      ,
   relative             ,
   makeShapeMarker      ,
   makeRangeMarker      ,
   makeFinancialMarker  ,
   makeLineMarker       ,
   makeLineStyle        ,

  -- * Setting Chart Parameters
   setChartSize                     ,
   setChartHeight                   ,
   setChartType                     ,
   setChartTitle                    ,
   setChartTitleWithColor           ,
   setChartTitleWithColorAndFontSize,
   setDataEncoding                  ,
   addChartData                     ,
   addDataScale                     ,
   addChartDataXY                   ,
   setColors                        ,
   addColor                         ,
   addFill                          ,
   setLegend                        ,
   addAxis                          ,
   setGrid                          ,
   addShapeMarker                   ,
   addRangeMarker                   ,
   addFinancialMarker               ,
   addLineMarker                    ,
   addLineFill                      ,
   setLabels                        ,
   setLabel                         ,
   setBarWidthSpacing               ,
   setPieChartOrientation           ,
   addLineStyle                     ,
   setFormula                       ,
   setQREncoding                    ,
   setQRWidth                       ,
   setQRErrorCorrection             ,
  -- * Retrieving Chart Data

  getChartData,
  getChartUrl,
  convertToUrl
)  where

import Graphics.GChart.Types
import Graphics.GChart.ChartItems
import Graphics.GChart.DataEncoding

import Data.List



-- | generates a 'Solid' fill from a hex color value
solid :: Color -> FillType -> Fill
solid color filltype = Fill (Solid color) filltype

-- | generates a 'ChartLegend' from a list of labels
legend :: [String] -> ChartLegend
legend labels = Legend labels Nothing

-- | generats a 'ChartLegend' from a list of lables and a 'LegendPosition'
legendWithPosition :: [String] -> LegendPosition -> ChartLegend
legendWithPosition labels position = Legend labels (Just position)

{-| returns a default axis. Use this to override the fields with your own
 values. For e.g :

@
makeAxis { 'axisType' = 'AxisTop',
           'axisLabels' = [\"0\",\"50\",\"100\"] }
@

-}
makeAxis :: Axis
makeAxis = defaultAxis


{-| returns a default axis. Use this to override the fields with your own
 values. For e.g :

@
makeGrid { 'xAxisStep' = 10,
           'yAxisStep' = 10,
            xOffset = Just 5 }
@

-}
makeGrid :: ChartGrid
makeGrid = defaultGrid

-- | Use this to specify the 'Simple' encoding for the 'setDataEncoding'
-- function.
simple :: ChartData
simple = Simple []

-- | Use this to specify the 'Text' encoding for the 'setDataEncoding'
-- function.
text :: ChartData
text = Text []

-- | Use this to specify the 'Extended' encoding for the 'setDataEncoding'
--  function.
extended :: ChartData
extended = Extended []

-- | Set automatic bar width for bar chart
automatic :: BarChartWidthSpacing
automatic = (Just Automatic,Nothing)

-- | Set automatic bar width for bar chart, with spacing values
automaticWithSpacing :: Int -> Int -> BarChartWidthSpacing
automaticWithSpacing b g= (Just Automatic, Just (Fixed (b,g)))

-- | Set bar width for chart
barwidth :: Int -> BarChartWidthSpacing
barwidth n = (Just (BarWidth n), Nothing)

-- | Set bar width and spacing for chart
barwidthspacing :: Int -> Int -> Int -> BarChartWidthSpacing
barwidthspacing bw b g = (Just (BarWidth bw), Just (Fixed (b,g)))

-- | Set relative spacing
relative :: Float -> Float -> BarChartWidthSpacing
relative b g = (Nothing, Just (Relative (b,g)))

-- | Shape Marker
makeShapeMarker :: ShapeMarker
makeShapeMarker = defaultShapeMarker

-- | Range Marker
makeRangeMarker :: RangeMarker
makeRangeMarker = defaultRangeMarker

-- | Line Marker
makeLineMarker :: LineMarker
makeLineMarker = defaultLineMarker

-- | Financial Marker
makeFinancialMarker :: FinancialMarker
makeFinancialMarker = defaultFinancialMarker

-- | Line Style
makeLineStyle :: LineStyle
makeLineStyle = defaultLineStyle

-- | Set the chart size by passing the width and the height in pixels
-- For e.g : @setChartSize 320 200@
setChartSize :: Int -> Int -> ChartM ()
setChartSize w h = set (Size w h)

-- | Set chart height only. Applicable to 'Formula' charts
-- This will set the width to 0 which will automatically
-- be excluded when the data is being encoded
setChartHeight :: Int -> ChartM ()
setChartHeight h = set (Size 0 h)

-- | Set the chart type by passing a 'ChartType'
setChartType :: ChartType -> ChartM ()
setChartType = set

-- | Set the chart title by passing a 'ChartTitle'
setChartTitle :: String -> ChartM ()
setChartTitle title = set $ ChartTitle title Nothing Nothing

-- | Set the chart title with a color
setChartTitleWithColor :: String -> Color -> ChartM()
setChartTitleWithColor title color = set $ ChartTitle title (Just color) Nothing

-- | Set the chart title with color and font size
setChartTitleWithColorAndFontSize :: String -> Color -> FontSize -> ChartM ()
setChartTitleWithColorAndFontSize title color fontsize =
    set $ ChartTitle title (Just color) (Just fontsize)

{-| Use this with 'simple', 'text' or 'extended' to specify the encoding. For e.g

@
setDataEncoding simple
@

Make sure you pass in values of the right type, Int for simple and extended
encoding, and Float for text encoding.
-}
setDataEncoding :: ChartData -> ChartM ()
setDataEncoding = set

{-| Add data to chart. Make sure you have set the data encoding using
 'setDataEncoding' before calling this function, otherwise it may generate
 gibberish, or throw an error
-}
addChartData :: ChartDataEncodable a => [a] -> ChartM ()
addChartData = addDataToChart


-- | Add a scale to chart.If more than one scale is added, it applies
-- the scale in order to each data series
addDataScale :: DataScale -> ChartM ()
addDataScale = addScaleToChart

-- | Works like 'addChartData', but for XY datasets for line XY chart etc
addChartDataXY :: ChartDataEncodable a => [(a,a)] -> ChartM ()
addChartDataXY series = do addDataToChart xseries
                           addDataToChart yseries
                        where xseries = map fst series
                              yseries = map snd series

-- | Pass a list of colors corresponding to the datasets in the chart
setColors :: [Color] -> ChartM ()
setColors = set . ChartColors

{-| Add a color to the chart. This color will be added to the list 'ChartColors'.

Make sure you do not include a call to 'setColors' at any time after a call to
'addColor', since this will lead to all previous values being erased.

-}
addColor :: Color -> ChartM()
addColor  = addColorToChart

-- | Add a 'Fill' to the chart
addFill :: Fill -> ChartM ()
addFill = addFillToChart

-- | Set a Legend for the chart
setLegend :: ChartLegend -> ChartM ()
setLegend = set

-- | Add an 'Axis' to the chart
addAxis :: Axis -> ChartM ()
addAxis = addAxisToChart

-- | Set a 'ChartGrid' for the chart
setGrid :: ChartGrid -> ChartM ()
setGrid = set

-- | Adds a shape marker. Use `makeShapeMarker` smart constructor when calling
-- this function If value of data set index is not specified when using
-- 'makeShapeMarker', it automatically adds a data index to refer to the latest
-- data set
addShapeMarker :: ShapeMarker -> ChartM ()
addShapeMarker marker | shapeDataSetIdx marker > (- 1) = addMarker marker
                      | otherwise = do idx <- getDataSetIdx
                                       let newmarker = marker { shapeDataSetIdx = idx }
                                       addMarker marker

-- | Adds a range marker. You can use 'makeRangeMarker' smart constructor when
-- calling this function
addRangeMarker :: RangeMarker -> ChartM ()
addRangeMarker = addMarker

-- | Adds a financial marker. Use 'makeFinancialMarker' smart constructor when
-- calling this function. If value of data set index is not specified when using
-- 'makeFinancialMarker', it automatically adds a data index to refer to the latest
-- data set
addFinancialMarker :: FinancialMarker -> ChartM ()
addFinancialMarker marker | financeDataSetIdx marker > (- 1) = addMarker marker
                          | otherwise = do idx <- getDataSetIdx
                                           let newmarker = marker { financeDataSetIdx = idx }
                                           addMarker newmarker
-- | Adds a line marker. Use 'makeLineMarker' smart constructor when calling
-- this function. If value of data set index is not specified when using
-- 'makeLineMarker', it automatically adds a data index to refer to the
-- latest data set
addLineMarker :: LineMarker -> ChartM ()
addLineMarker marker | lineDataSetIdx marker > (- 1) = addMarker marker
                     | otherwise = do idx <- getDataSetIdx
                                      let newmarker = marker { lineDataSetIdx = idx }
                                      addMarker newmarker

-- | Adds a line fill to the chart
addLineFill :: LineFillType -> Color -> ChartM ()
addLineFill fillType color = addMarker (LineFillMarker fillType color)

-- | Set labels for the chart
setLabels :: [String] -> ChartM ()
setLabels = set . ChartLabels

-- | Set label for a chart
setLabel :: String -> ChartM ()
setLabel label = set $ ChartLabels [label]

-- | Set bar and width spacing
setBarWidthSpacing :: BarChartWidthSpacing -> ChartM ()
setBarWidthSpacing = set

-- | Set pie chart orientation in radians
setPieChartOrientation :: Float -> ChartM ()
setPieChartOrientation = set . PCO

-- | Add line style
addLineStyle :: LineStyle -> ChartM()
addLineStyle = addLineStyleToChart

-- | Set formula. Applies only to 'Formula' charts
setFormula :: String -> ChartM ()
setFormula formula = setLabels [formula]

-- | Set QR code output encoding. Valid for 'QRCode' only
setQREncoding :: QREncoding -> ChartM ()
setQREncoding = set

-- | Sets the error correction level for 'QRCode'
setQRErrorCorrection :: ErrorCorrectionLevel -> ChartM ()
setQRErrorCorrection ec = set qrLabelData where
                           qrLabelData = QRLabelData ec defMargin
                           QRLabelData _ defMargin = defaultQREncodingLabelData

-- | Sets the width (in rows) of the white border around the data portion of the 'QRCode'
setQRWidth :: Int -> ChartM ()
setQRWidth m = set qrLabelData where
                qrLabelData = QRLabelData defEC m
                QRLabelData defEC _ = defaultQREncodingLabelData

-- | Extracts the data out of the monad and returns a value of type 'Chart'
getChartData :: ChartM () -> Chart
getChartData = getChartDataFromChartM

-- | Extracts the data out of the monad and returns a URL string for the chart
getChartUrl :: ChartM () -> String
getChartUrl =  convertToUrl . getChartData

-- | Converts a value of type 'Chart' to a URL
convertToUrl :: Chart -> String
convertToUrl chart = baseURL ++ intercalate "&" urlparams where
    baseURL = "http://chart.apis.google.com/chart?"
    urlparams = [urlEnc a ++ "=" ++ urlEnc b | (a,b) <- getParams chart]


Graphics/GChart/ChartItems/Colors.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE TypeSynonymInstances #-}
module Graphics.GChart.ChartItems.Colors where

import Graphics.GChart.Types
import Graphics.GChart.ChartItems.Util

import Data.List(intercalate)
-- Chart Colors
instance ChartItem ChartColors where
    set colors = updateChart $ \chart -> chart { chartColors = Just colors }

    encode (ChartColors colors) = asList ("chco", intercalate "," colors)

-- TODO: Fill Area

-- Chart Fills
instance ChartItem ChartFills where
    set fills = updateChart $ \chart -> chart { chartFills = Just fills }

    encode fills = asList ("chf",intercalate "|" $ map encodeFill fills)


encodeFill (Fill kind fType) = case kind of
                                 Solid color -> intercalate "," [fillType,"s",color]

                                 LinearGradient angle offsets -> intercalate "," [fillType,
                                                                                  "lg",
                                                                                  show angle,
                                                                                  intercalate "," $ map  (\(c,o) -> c ++ "," ++ show o ) offsets]

                                 LinearStripes angle widths ->   intercalate "," [fillType,
                                                                                  "ls",
                                                                                  show angle,
                                                                                  intercalate "," $ map (\(c,w) -> c ++ "," ++ show w) widths]
                                 where fillType = case fType of
                                                    Background  -> "bg"
                                                    Area        -> "c"
                                                    Transparent -> "a"


examples/Examples.hs
import Graphics.GChart
import Data.Char(ord)
{-

Examples to demonstrate usage of GChart.

Some examples are taken from this article : http://24ways.org/2007/tracking-christmas-cheer-with-google-charts

-}

christmasPie = getChartUrl $ do setChartSize 600 300
                                setDataEncoding simple
                                setChartType Pie
                                addChartData dataSeries1
                                setChartTitle "Food and Drink Consumed Christmas 2007"
                                setLabels labelSeries1
                                setColors ["00AF33",
                                           "4BB74C",
                                           "EE2C2C",
                                           "CC3232",
                                           "33FF33",
                                           "66FF66",
                                           "9AFF9A",
                                           "C1FFC1",
                                           "CCFFCC" ]


barGraph = getChartUrl $ do setChartSize 600 300
                            setDataEncoding simple
                            setChartType BarHorizontalGrouped
                            addChartData dataSeries1
                            setChartTitle "Food and Drink Consumed Christmas 2007"
                            addAxis $ makeAxis {  axisType = AxisBottom }
                            addAxis $ makeAxis {  axisType = AxisLeft,
                                                  axisLabels = Just labelSeries1 }
                            addColor "00AF33"

linexyGraph1 =
    getChartUrl $ do setChartSize 800 300
                     setChartType LineXY
                     setDataEncoding text
                     setChartTitle "Projected Christmas Cheer for 2007"
                     setGrid $ makeGrid { xAxisStep = 3.333,
                                          yAxisStep = 10,
                                          lineSegmentLength = Just 1,
                                          blankSegmentLength = Just 3 }
                     addAxis $ makeAxis { axisType = AxisLeft,
                                          axisRange = Just $ Range (0,100) (Just 50) }
                     addAxis $ makeAxis { axisType = AxisBottom,
                                          axisLabels = Just $ ["Dec 1st"] ++ blanks 4 ++ ["6th"] ++ blanks 18 ++ ["25th","26th"] ++ blanks 4 ++ ["Dec 31st"] }
                     addChartDataXY dataSeries2

linexyGraph2 =
    getChartUrl $ do setChartSize 800 300
                     setChartType LineXY
                     setDataEncoding text
                     setChartTitle "Projected Christmas Cheer for 2007"

                     setGrid $ makeGrid { xAxisStep = 3.333,
                                          yAxisStep = 10,
                                          lineSegmentLength = Just 1,
                                          blankSegmentLength = Just 3 }

                     addAxis $ makeAxis { axisType = AxisLeft,
                                          axisRange = Just $ Range (0,100) (Just 50) }
                     addAxis $ makeAxis { axisType = AxisBottom,
                                          axisLabels = Just $ ["Dec 1st"] ++ blanks 4 ++ ["6th"] ++ blanks 18 ++ ["25th","26th"] ++ blanks 4 ++ ["Dec 31st"] }

                     addChartDataXY dataSeries3
                     addColor "458B00"

                     addChartDataXY dataSeries4
                     addColor "CD2626"

                     setLegend $ legendWithPosition ["2006","2007"] LegendRight

bargraph2 = getChartUrl $ do setChartSize 600 300
                             setChartType BarHorizontalGrouped
                             setDataEncoding simple
                             addChartData dataSeries1
                             setChartTitleWithColor "Food and Drink Consumed Christmas 2007" "00AF33"
                             addAxis $ makeAxis {  axisType = AxisBottom }
                             addAxis $ makeAxis {  axisType = AxisLeft,
                                                   axisLabels = Just labelSeries1 }
                             addColor "00AF33"


bargraphAutoSpacing = getChartUrl $ do setChartSize 190 125
                                       setChartType BarVerticalGrouped
                                       setDataEncoding simple
                                       addChartData ([10,15,20,25,30]::[Int])
                                       addChartData ([13,5,6,34,12]::[Int])
                                       setColors ["4d89f9","000000"]
                                       setBarWidthSpacing automatic


bargraphRelativeSpacing = getChartUrl $ do setChartSize 190 125
                                           setChartType BarVerticalGrouped
                                           setDataEncoding simple
                                           addChartData ([10,15,20,25,30]::[Int])
                                           addChartData ([13,5,6,34,12]::[Int])
                                           setColors ["4d89f9","000000"]
                                           setBarWidthSpacing $ relative 0.5 1.5


scatterPlotWithMarkers = getChartUrl $ do setChartSize 200 125
                                          setChartType ScatterPlot
                                          setDataEncoding simple
                                          addChartDataXY dataSeries5
                                          addAxis $ makeAxis { axisType = AxisBottom,
                                                               axisLabels = Just $ blanks 1 ++ ["1","2","3","4","5"] }
                                          addAxis $ makeAxis { axisType = AxisLeft,
                                                               axisLabels = Just $ blanks 1 ++ ["50","100"] }
                                          setGrid $ makeGrid { xAxisStep = 20, yAxisStep = 25 }

                                          addShapeMarker $ makeShapeMarker { shapeType  = ShapeSquare
                                                                           , shapeColor = "ff0000"
                                                                           , shapeSize  = 10 }

lineChartWithLineStyles = getChartUrl $ do setChartSize 200 125
                                           setChartType Line
                                           setDataEncoding simple

                                           addChartData $ map encSimpleReverse "93zyvneTTOMJMLIJFHEAECFJGHDBFCFIERcgnpy45879"
                                           addLineStyle $ makeLineStyle { lineStyleThickness = 3, lineStyleLineSegment = 6, lineStyleBlankSegment = 3 }

                                           addChartData $ map encSimpleReverse "IJKNUWUWYdnswz047977315533zy1246872tnkgcaZQONHCECAAAAEII"
                                           addLineStyle $ makeLineStyle { lineStyleThickness = 1, lineStyleLineSegment = 1, lineStyleBlankSegment = 0 }

formulaChart = getChartUrl $ do setChartType Formula
                                setChartHeight 200
                                setFormula "\\Large\\mathbb{Q}+\\lim_{x\\to3}\\frac{1}{x}"
                                addColor "FF0000"
                                addFill $ Fill (LinearGradient 20 [("76A4FB",1),("FFFFFF",0)]) Background

qrCodeChart = getChartUrl $ do setChartType QRCode
                               setChartSize 150 150
                               setLabel "Hello World"
                               setQREncoding UTF8
                               setQRErrorCorrection L'

barGraphWithShiftedZeroLine = getChartUrl $ do setChartType BarVerticalGrouped
                                               setChartSize 320 200
                                               setDataEncoding text
                                               addChartData ([30,-60,50,140,80,-90]::[Float])
                                               addDataScale (-80,140)
                                               addAxis $ makeAxis { axisType = AxisLeft,
                                                                    axisRange =  Just $ Range (-80,140) Nothing}

lineMarkerSample = getChartUrl $ do setChartType BarVerticalGrouped
                                    setChartSize 200 150
                                    setDataEncoding simple
                                    setBarWidthSpacing $ barwidth 20
                                    addChartData $ map encSimpleReverse "1XQbnf4"
                                    addColor "76A4FB"
                                    addLineMarker $ makeLineMarker { lineColor = "0033FF",
                                                                     lineWhichPoints = PointsAll,
                                                                     lineZorder = 1 }

lineFillSample = getChartUrl $ do setChartType Line
                                  setChartSize 200 125
                                  setDataEncoding simple
                                  addChartData $ map encSimpleReverse "cefhjkqwrlgYcfgc"
                                  addLineFill (LineFillBetween 0 1) "224499"
                                  addChartData $ map encSimpleReverse "QSSVXXdkfZUMRTUQ"
                                  addLineFill (LineFillBetween 1 2) "FF0000"
                                  addChartData $ map encSimpleReverse "HJJMOOUbVPKDHKLH"
                                  addLineFill (LineFillBetween 2 3) "80C65A"
                                  addAxis $  makeAxis { axisType = AxisBottom,
                                                       axisLabels = Just  ["Sep","Oct","Nov","Dec"] }
                                  addAxis $ makeAxis { axisType = AxisLeft,
                                                       axisLabels = Just  ["","50","100"] }

blanks x = replicate x ""

dataSeries1 :: [Int]
dataSeries1 = [10,20,8,25,5,3,15,9,5]

dataSeries2 :: [(Float,Float)]
dataSeries2 = [(0,0),(100,100)]

dataSeries3 :: [(Float,Float)]
dataSeries3 = zip [0,16.7,23.3,33.3,60,76.7,83.3,86.7,93.3,96.7,100] [30,45,20,50,15,80,60,70,40,55,80]

dataSeries4 :: [(Float,Float)]
dataSeries4 = zip [0,10,16.7,26.7,33.3] [50,10,30,55,60]

dataSeries5 :: [(Int,Int)]
dataSeries5 = zip xseries yseries where
             xseries = map encSimpleReverse "984sttvuvkQIBLKNCAIipr3z9"
             yseries = map encSimpleReverse "DEJPgq0uov17_zwopQOD"

labelSeries1 = ["Egg nog",
                "Christmas Ham",
                "Milk (not including egg nog)",
                "Cookies",
                "Roasted Chestnuts",
                "Chocolate",
                "Various Other Beverages",
                "Various Other Foods",
                "Snacks"]

encSimpleReverse :: Char -> Int
encSimpleReverse c | ord c >= ord 'A' && ord c <= ord 'Z' = ord c - ord 'A'
                   | ord c >= ord 'a' && ord c <= ord 'z' = 26 + (ord c - ord 'a')
                   | ord c >= ord '0' && ord c <= ord '9' = 52 + (ord c - ord '0')
                   | otherwise = -1

main = do putStrLn christmasPie
          putStrLn barGraph
          putStrLn linexyGraph1
          putStrLn linexyGraph2
          putStrLn bargraph2
          putStrLn bargraphAutoSpacing
          putStrLn bargraphRelativeSpacing
          putStrLn scatterPlotWithMarkers
          putStrLn lineChartWithLineStyles
          putStrLn formulaChart
          putStrLn qrCodeChart
